src/main/scala/analysis/DSAUtility.scala

```diff
@@ -0,0 +1,983 @@
+
+  override def toString: String = s"Node($id, $allocationRegions ${if collapsed then ", collapsed" else ""})"
+
+
+}
+
+/**
+ * a cell in DSA
+ * @param node the node this cell belongs to
+ * @param offset the offset of the cell
+ */
+case class DSC(node: Option[DSN], offset: BigInt)
+{
+  var largestAccessedSize: BigInt = 0
+
+  var _pointee : Option[Slice] = None
+
+  def getPointee : Slice =
+    if _pointee.isEmpty then
+      val node = DSN(Some(this.node.get.graph.get))
+      _pointee = Some(Slice(node.cells(0), 0))
+    else
+
+      val graph = _pointee.get.node.graph.get
+      val resolvedPointee = graph.find(graph.adjust(_pointee.get))
+
+      _pointee = Some(graph.deadjust(resolvedPointee))
+    _pointee.get
+
+  def growSize(size: BigInt): Boolean =
+    if size > largestAccessedSize then
+      largestAccessedSize = size
+      true
+    else false
+
+  override def toString: String = s"Cell(${if node.isDefined then node.get.toString else "NONE"}, $offset)"
+}
+
+
+/**
+ * a slice made from a cell and an internal offset
+ */
+case class Slice(cell: DSC, internalOffset: BigInt) {
+  def node: DSN  = cell.node.get
+  def offset: BigInt = cell.offset
+}
+
+/**
+ * represents a direct call in DSA
+ * @param call instance of the call
+ * @param graph caller's DSG
+ */
+class CallSite(val call: DirectCall, val graph: DSG) {
+  val proc = call.target
+  val paramCells: mutable.Map[Variable, Slice] = graph.params(proc).foldLeft(mutable.Map[Variable, Slice]()) {
+    (m, reg) =>
+      val node = DSN(Some(graph))
+      node.flags.incomplete = true
+      m += (reg -> Slice(node.cells(0), 0))
+  }
+  val returnCells: mutable.Map[Variable, Slice] = graph.writesTo(proc).foldLeft(mutable.Map[Variable, Slice]()) {
+    (m, reg) =>
+      val node = DSN(Some(graph))
+      node.flags.incomplete = true
+      m += (reg -> Slice(node.cells(0), 0))
+  }
+}
+
+// global address range
+case class AddressRange(start: BigInt, end: BigInt)
+
+// a node, offset pair, difference to a cell is that it doesn't represent a DSG construct,
+case class Field(node: DSN, offset: BigInt)
+
+// unwraps internal padding and slicing and returns the expression
+def unwrapPaddingAndSlicing(expr: Expr): Expr =
+  expr match
+    case literal: Literal => literal
+    case Repeat(repeats, body) => Repeat(repeats, unwrapPaddingAndSlicing(body))
+    case SignExtend(extension, body) => SignExtend(extension, unwrapPaddingAndSlicing(body))
+    case UnaryExpr(op, arg) => UnaryExpr(op, arg)
+    case BinaryExpr(op, arg1, arg2) => BinaryExpr(op, unwrapPaddingAndSlicing(arg1), unwrapPaddingAndSlicing(arg2))
+    case MemoryLoad(mem, index, endian, size) => MemoryLoad(mem, unwrapPaddingAndSlicing(index), endian, size)
+    case variable: Variable => variable
+    case Extract(end, start, body) /*if start == 0 && end == 32*/ => unwrapPaddingAndSlicing(body) // this may make it unsound
+    case ZeroExtend(extension, body) => unwrapPaddingAndSlicing(body)
+    case _ => expr
```

Comment:
Ignoring extract and extend operations is not going to produce correct results.

Reply:
Since pointers are 64-bit values, if a SignExtend/ZeroExtend/Repeat/Extract is applied to a pointer, the result will not be a  pointer, so I assume that means the nodes should be collapsed? It may be possible to refine that though, some of the cases we need to care about don't exist in higher level code so they weren't considered in the DSA paper.